src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -4624,6 +4632,283 @@ std::unique_ptr<LCSerializer> RadosLifecycle::get_serializer(const std::string&
+    bufferlist bl;
+
+    encode(entry, bl);
+    ent_list.push_back(std::move(bl));
+
+  }
+
+  ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__
+		 << "Adding " << restore_entries.size() << " entries to FIFO:" << obj_names[index] << dendl;
+
+  int ret = push(dpp, y, index, std::move(ent_list));
+
+  if (ret < 0) {
+    ldpp_dout(dpp, -1) << "ERROR: push() returned " << ret << dendl;
+    return ret;
+  }
+
+  return 0;
+}
+
+int RadosRestore::push(const DoutPrefixProvider *dpp, optional_yield y,
+		int index, std::vector<ceph::buffer::list>&& items) {
+  ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__
+		 << "Pushing entries to FIFO:" << obj_names[index] << dendl;
+  try {
+    fifos[index]->push(dpp, items, ceph::async::use_blocked);
+  } catch (const sys::system_error& e) {
+    ldpp_dout(dpp, -1) << __PRETTY_FUNCTION__
+		 << ": unable to push to FIFO: " << obj_names[index]
+		 << ": " << e.what() << dendl;
+    throw;
+  }
+  return 0;
+}
+
+int RadosRestore::push(const DoutPrefixProvider *dpp, optional_yield y,
+		int index, ceph::buffer::list&& bl) {
+  ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__
+		 << "Pushing entry to FIFO:" << obj_names[index] << dendl;
+
+  try {
+    fifos[index]->push(dpp, std::move(bl), ceph::async::use_blocked);
+  } catch (const sys::system_error& e) {
+    ldpp_dout(dpp, -1) << __PRETTY_FUNCTION__
+		 << ": unable to push to FIFO: " << obj_names[index]
+		 << ": " << e.what() << dendl;
+    throw;
+  }
+  return 0;
+}
+
+struct rgw_restore_fifo_entry {
+  std::string id;
+  ceph::real_time mtime;
+  RGWRestoreEntry entry;
+  rgw_restore_fifo_entry() {}
+
+  void encode(ceph::buffer::list& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(id, bl);
+    encode(mtime, bl);
+    encode(entry, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(ceph::buffer::list::const_iterator& bl) {
+     DECODE_START(1, bl);
+     decode(id, bl);
+     decode(mtime, bl);
+     decode(entry, bl);
+     DECODE_FINISH(bl);
+  }
+
+  void dump(ceph::Formatter* f) const;
+  void decode_json(JSONObj* obj);
+};
+WRITE_CLASS_ENCODER(rgw_restore_fifo_entry)
+
+int RadosRestore::list(const DoutPrefixProvider *dpp, optional_yield y,
+	       	   int index, const std::string& marker, std::string* out_marker,
+		   uint32_t max_entries, std::vector<RGWRestoreEntry>& entries,
+		   bool* truncated)
+{
+  std::vector<fifo::entry> restore_entries{max_entries};
+  std::string omark = {};
+  bool more = false;
+
+  ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__
+		 << "Listing entries from FIFO:" << obj_names[index] << dendl;
+
+  try {
+    auto [lentries, lmark] = fifos[index]->list(dpp, marker,
+			  	 restore_entries, ceph::async::use_blocked);
+    entries.clear();
+
+    for (const auto& entry : lentries) {
+      rgw_restore_fifo_entry r_entry;
+      r_entry.id = entry.marker;
+      r_entry.mtime = entry.mtime;
+
+      auto liter = entry.data.cbegin();
+      try {
+	decode(r_entry.entry, liter);
+      } catch (const buffer::error& err) {
```

Comment:
You should be able to just move this catch down above the other one and not need the extra try block.

Reply:
Done.